File: governance/ui/dashboard_server.py
"""
PMO Dashboard Server
Serves the dashboard UI and handles API calls (clear history).
"""

import http.server
import json
import shutil
from pathlib import Path
from datetime import datetime
import os

PORT = 8765
UI_DIR = Path(__file__).parent
DATA_DIR = Path(__file__).parent.parent / "data"
STATE_FILE = DATA_DIR / "collab_state.json"
MESSAGE_LOG = DATA_DIR / "collab_messages.jsonl"
ARCHIVE_DIR = DATA_DIR / "archive"


class Handler(http.server.SimpleHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/" or self.path == "/ui/pmo_dashboard.html":
            self.path = "/ui/pmo_dashboard.html"

        if self.path.startswith("/ui/"):
            # Serve from UI directory
            file_path = UI_DIR / self.path[1:]
            if file_path.exists() and file_path.is_file():
                self.send_file(file_path)
                return

        if self.path == "/api/collabs" or self.path == "/data/collab_state.json":
            self.send_json(json.loads(STATE_FILE.read_text()))
            return

        if self.path == "/api/messages" or self.path == "/data/collab_messages.jsonl":
            text = MESSAGE_LOG.read_text() if MESSAGE_LOG.exists() else ""
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", len(text))
            self.end_headers()
            self.wfile.write(text.encode())
            return

        # Fallback: serve from ui dir
        if self.path == "/pmo_dashboard.html":
            self.send_file(UI_DIR / "pmo_dashboard.html")
            return

        self.send_error(404)

    def do_POST(self):
        if self.path == "/api/clear-history":
            ARCHIVE_DIR.mkdir(exist_ok=True)
            ts = datetime.now().strftime("%Y%m%d-%H%M%S")

            shutil.copy(STATE_FILE, ARCHIVE_DIR / f"collab_state_{ts}.json")
            if MESSAGE_LOG.exists():
                shutil.copy(MESSAGE_LOG, ARCHIVE_DIR / f"collab_messages_{ts}.jsonl")

            STATE_FILE.write_text("{}")
            MESSAGE_LOG.write_text("")

            self.send_json({"ok": True, "archived_to": str(ARCHIVE_DIR)})
            return

        self.send_error(404)

    def send_file(self, file_path):
        ctype = "text/html" if file_path.suffix == ".html" else "text/plain"
        with open(file_path, "rb") as f:
            data = f.read()
        self.send_response(200)
        self.send_header("Content-Type", ctype)
        self.send_header("Content-Length", len(data))
        self.end_headers()
        self.wfile.write(data)

    def send_json(self, data):
        text = json.dumps(data)
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", len(text))
        self.end_headers()
        self.wfile.write(text.encode())

    def log_message(self, fmt, *args):
        pass


if __name__ == "__main__":
    server = http.server.HTTPServer(("0.0.0.0", PORT), Handler)
    print(f"PMO Dashboard: http://localhost:{PORT}/ui/pmo_dashboard.html")
    server.serve_forever()

File: governance/ui/dashboard_server.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const querystring = require('querystring');

const PORT = 8765;
const DATA_DIR = path.join(__dirname, '..', 'data');
const STATE_FILE = path.join(DATA_DIR, 'collab_state.json');
const MSG_FILE = path.join(DATA_DIR, 'collab_messages.jsonl');
const ARCHIVE_DIR = path.join(DATA_DIR, 'archive');

const MIME = { '.html': 'text/html', '.json': 'application/json', '.js': 'text/javascript' };

const server = http.createServer((req, res) => {
    const url = req.url.split('?')[0];

    if (req.method === 'GET') {
        // Dashboard
        if (url === '/ui/pmo_dashboard.html' || url === '/') {
            const file = path.join(__dirname, 'pmo_dashboard.html');
            fs.readFile(file, (err, data) => {
                if (err) { res.writeHead(404); res.end('404'); return; }
                res.writeHead(200, { 'Content-Type': 'text/html' });
                res.end(data);
            });
            return;
        }

        // API: collabs state
        if (url === '/api/collabs' || url === '/data/collab_state.json') {
            try {
                const data = fs.existsSync(STATE_FILE) ? fs.readFileSync(STATE_FILE, 'utf8') : '{}';
                res.writeHead(200, { 'Content-Type': 'application/json' });
                res.end(data);
            } catch(e) { res.writeHead(500); res.end('error'); }
            return;
        }

        // API: messages
        if (url === '/api/messages' || url === '/data/collab_messages.jsonl') {
            try {
                const data = fs.existsSync(MSG_FILE) ? fs.readFileSync(MSG_FILE, 'utf8') : '';
                res.writeHead(200, { 'Content-Type': 'application/json' });
                res.end(data);
            } catch(e) { res.writeHead(500); res.end('error'); }
            return;
        }

        res.writeHead(404);
        res.end('404');
        return;
    }

    if (req.method === 'POST' && url === '/api/clear-history') {
        try {
            fs.mkdirSync(ARCHIVE_DIR, { recursive: true });
            const ts = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
            if (fs.existsSync(STATE_FILE)) fs.copyFileSync(STATE_FILE, path.join(ARCHIVE_DIR, `collab_state_${ts}.json`));
            if (fs.existsSync(MSG_FILE)) fs.copyFileSync(MSG_FILE, path.join(ARCHIVE_DIR, `collab_messages_${ts}.jsonl`));
            fs.writeFileSync(STATE_FILE, '{}');
            fs.writeFileSync(MSG_FILE, '');
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ ok: true, archived_to: ARCHIVE_DIR }));
        } catch(e) {
            res.writeHead(500, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ error: e.message }));
        }
        return;
    }

    res.writeHead(404);
    res.end('404');
});

server.listen(PORT, () => console.log(`PMO Dashboard: http://localhost:${PORT}/ui/pmo_dashboard.html`));